textx/lang.py

```diff
@@ -535,6 +535,12 @@ def visit_rule_name(self, node, children):
             self.dprint("Creating class: {}".format(rule_name))
 
         # If a class is given by the user use it. Else, create new class.
+        if rule_name not in self.metamodel.user_classes:
+            user_class = self.metamodel.user_classes_provider(
+                rule_name
+            )
+            if user_class is not None:
+                self.metamodel.user_classes[rule_name] = user_class
```

Comment:
This block of code might be reworked to improve perf., something like this (not tested):
```python
         if self.metamodel.user_classes_provider is not None:
            cls = self.metamodel.user_classes_provider(
                rule_name
            )
            if cls is not None:
                self.metamodel.user_classes[rule_name] = cls
         else:
             cls = self.metamodel.user_classes.get(rule_name)

         if cls is not None:
            # Initialize special attributes
            self.metamodel._init_class(cls, None, node.position,
                                       external_attributes=True)
         else:
            # Create class to collect attributes. At this time PEG rule
            # is not known.
            cls = self.metamodel._new_class(rule_name, None, node.position)

```

Reply:
ok. added an assert.